pow-migration/src/history/mod.rs

```diff
@@ -102,100 +105,128 @@ fn from_pow_transaction(pow_transaction: &PoWTransaction) -> Result<Transaction,
-    for block_height in start..cutting_pow_block_number {
-        // Refresh the progress bar position
-        pb.set_position(block_height as u64);
+        log::info!(
+            candidate_block,
+            "Start migrating PoW history up to the next candidate block"
+        );
 
-        // Get all transactions for this block height
-        let mut transactions = vec![];
-        let block = client.get_block_by_number(block_height, false).await?;
-        let mut network_id = NetworkId::Main;
-        match block.transactions {
-            PoWTransactionSequence::BlockHashes(hashes) => {
-                if hashes.is_empty() {
-                    continue;
-                }
-                for hash in hashes {
-                    log::trace!(hash, "Processing transaction");
-                    let pow_transaction = client.get_transaction_by_hash_2(&hash).await?;
-                    let pos_transaction = from_pow_transaction(&pow_transaction)?;
-                    network_id = pos_transaction.network_id;
+        // Get transactions of each block and add them to the PoS history store
+        for block_height in (history_store_height + 1)..=candidate_block {
+            // Check if we are closing in on the PoW head block.
+            // If this is true, we may need to take some extra time in order to make sure
+            // that blocks leading up to the head block are confirmed before we
+            // can safely add them to the PoS history store.
+            if pow_head_height - block_height <= block_confirmations {
+                loop {
+                    // Check if the block has been confirmed.
+                    if pow_head_height > block_height + block_confirmations {
+                        break;
+                    }
 
-                    assert_eq!(
-                        pow_transaction.hash,
-                        pos_transaction.hash::<Blake2bHash>().to_hex()
+                    pow_head_height = pow_client.block_number().await.unwrap();
+                    log::info!(
+                        "Waiting for block {} to be confirmed before we can migrate it",
+                        block_height
                     );
-                    transactions.push(ExecutedTransaction::Ok(pos_transaction));
+                    sleep(Duration::from_secs(60)).await;
                 }
             }
-            PoWTransactionSequence::Transactions(_) => panic!("Unexpected transaction type"),
+
+            let block = pow_client
+                .get_block_by_number(block_height, false)
+                .await
+                .unwrap();
+
+            // Get all transactions for this block height
+            let mut transactions = vec![];
+            let mut network_id = NetworkId::Main;
+            match block.transactions {
+                PoWTransactionSequence::BlockHashes(hashes) => {
+                    // Mark that we've migrated up until this block.
```

Comment:
I'd do this after we actually have done the migration for all of the hashes of this block height